sxs/waveforms/alignment.py

```diff
@@ -310,3 +299,227 @@ def align2d(wa, wb, t1, t2, n_brute_force_δt=None, n_brute_force_δϕ=5, includ
     idx = np.argmin(abs(np.array([optimum.cost for optimum in optimums])))
 
     return optimums[idx].cost, wa_primes[idx], optimums[idx]
+
+
+def _cost4d(δt_δSO3, args):
+    from .. import WaveformModes
+
+    modes_A, modes_B, t_reference, normalization = args
+    δt = δt_δSO3[0]
+    δSO3 = np.exp(quaternion.quaternion(*δt_δSO3[1:]))
+
+    modes_A_at_δt = modes_A(t_reference + δt)
+    ell_max = int(np.sqrt(modes_A_at_δt.shape[1] + 4)) - 1
+
+    wa_prime = WaveformModes(
+        input_array=(modes_A_at_δt),
+        time=t_reference,
+        time_axis=0,
+        modes_axis=1,
+        ell_min=2,
+        ell_max=ell_max,
+        spin_weight=-2,
+    )
+
+    wa_prime = wa_prime.rotate(δSO3.components)
+
+    # Take the sqrt because least_squares squares the inputs...
+    diff = trapezoid(
+        np.sum(abs(wa_prime.data - modes_B) ** 2, axis=1),
+        t_reference,
+    )
+    return np.sqrt(diff / normalization)
```

Comment:
If you're passing this to `least_squares`, you should just return the residuals, rather than summing or integrating them — and certainly not the sqrt.  Normalization is fine, and you'll probably want to reinterpret the complex difference with `.view(float)`.

Reply:
I was just repeating the style of https://github.com/sxs-collaboration/sxs/blob/0b5182768956dced348173d7fad7951a6e751e80/sxs/waveforms/alignment.py#L109 and https://github.com/sxs-collaboration/sxs/blob/0b5182768956dced348173d7fad7951a6e751e80/sxs/waveforms/alignment.py#L123. Do you want to change those as well?